Comprehensive Queries

Yanrishatum opened the pull request:
Hoo boy, this is a big one.
Introduces an overhauled SelectorQuery system where content can be labeled and use separate queries to find it, allowing far more precise control over the content as well as lets us construct more robust queries.

# Overview
* SelectorQuery changes
  * Removed Parcelable from it as it's a dead weight.
  * Added `host:String?` - extra filter to match only specific hosts.
  * Added `keepContentStyle/Ids/Classes` - Query now directly controls if any of those have to be removed or not.
  * Added `customCSS:String` - Allows to add custom CSS for query. Allows us to fine-tune specific websites display state. (Like activetranslations)
  * Added a subquery system (see below)
* Added SelectorSubquery, refer to SelectorSubquery and SubqueryRole for their documentation.
* Deprecated legacy tumblr cleaner in favor of comprehensive selectors.
* Deprecated custom activetranslations cleaner in favor of comprehensive selector.
* Deprecated circustranslations cleaner because website is dead and even domain name is up for sale.
* Changed the ordering of htmlCleanerQueries, user-defined ones are now prioritised.
* Added github .md file query
* Added `article` query as a last resort.

# How it works
1. The matching query now also checks for host match, meaning that it's important to put host-restricted queries before unrestricted ones, just so the search would check them first.
2. When host match is found (or host is null), regular query match is performed.
3. If main query is found, all non-optional subqueries are checked.
4. If all of the above pass, the query is deemed correct and thus used by cleaner.
5. 1. Cleaner uses legacy behavior if query does not contain any subqueries and uses primary query as content, skip to step 6.
5. 2. Otherwise it fetches the matches of each subquery and goes trough them, constructing cleaned up content.
5. 3. Depending on subquery role - match is either discarded, processed or transformed. Unless it is removed (or whitelisted) content will always get CSS-cleaned.
6. If no Header subquery role was present or that role ended up being optional and didn't yield results, depending on appendTitleHeader it will add the page title as a replacement.
7. After the cleanup operation is done, cleaned content is used.
8. If present, custom css is injected.

# Considerations
I'm not sure if SelectorQuery should contain equals/hashCode and toString, as they should be derived by compiler (since it's data class)? And if they have to have it - I'm not sure I implemented it properly. :) Also same question goes to SelectorSubquery.

To avoid double-querying of same stuff, RContent subquery can contain an empty string for a query, which is then interpreted as "use primary query here" (same happens if it's optional and doesn't yield any results)

Also see TODOs about some optimisations and changes that may be done.

I'm not sure how serialization works on List<T>. But in general it's schema is pretty simple:
```ts
class SelectorQuery {
  query: String,
  appendTitleHeader: Boolean = true,
  host: String = null,
   subqueries:Array<Subquery> = [];
}

class Subquery {
  query: String,
  role: SubqueryRole,
  optional: Boolean = true,
  multiple: Boolean = true
}

enum SubqueryRole {
    RContent, RHeader, RFooter, RShare,
    RComments, RMeta, RNavigation,
    RPage, RBlacklist, RWhitelist
}
```
Note that only element that is allowed to be null is `host`. Others have to be non-null, including subqueries (as an empty array)

gmathi commented:
@Yanrishatum Couple+1 of suggestions. 

1. I read up on if we need the serialization for these `data` classes. And the answer is no. We can remove the `Serializable` implementation. 
2. All of the data classes can be moved into one `HtmlSelectorModels.kt` since they are one-line definitions anyway. 
3. `SelectorQuery` & `SelectorSubQuery` can inherit the `Query` base data class which contains `var query: String` for now but can be used in the future if we need to improvise the query class itself. (Optional)
4. Rename `query` variable to `selector`? Since `query.selector` is better than `query.query`? I think... I will leave it to your choice.

Yanrishatum commented:
1. Removed.
2. Moved.
3. Not sure how well data classes handle inheritance, plus it'll require code changes on the app side in any case, so I think it's overengineering at this point until actual need to do so arises (and if it will - separating some values would be simple anyway)
4. Renamed.

Also changed some checks to use `isEmpty` / `isNotEmpty` methods instead of comparing `count()` (not used to kotlin having those)
Also removed the unused cleaner files as to reduce clutter a bit. Ideally comprehensive queries should have enough features to cover most of custom cleaners currently used.
